### ISSUE TITLE
Panic in definition list support

### ISSUE THREAD

**Issue by ssokolow** [pulldown-cmark/pulldown-cmark#949]
When I upgraded pulldown-cmark to 0.12.1 in my little frontend and tried to enable definition list support, I got a panic at `src/firstpass.rs:1507:38` from the markdown file that acts as a mix of syntax reference and test suite.

I narrowed it down to this chunk of [Prettier](https://prettier.io/)-autoformatted text:

```markdown
**Example:**

> The Konami Code originally appeared in
> [Gradius (1986)](https://www.mobygames.com/game/4512/gradius/) on the NES,
> where it took the form
> <span style="white-space: nowrap">:dpad-up::dpad-up::dpad-down::dpad-down::dpad-left::dpad-right::dpad-left::dpad-right::btn-b-nes::btn-a-nes:.</span>
> Rather than being part of the code, the addition of :btn-start-nes: and/or
> :btn-select-nes: is actually part of selecting the number of players and
> starting a new game _after_ the code has been entered. (Naturally, that makes
> easter eggs difficult on Sony consoles, which have
> :btn-triangle:,:btn-circle:,:btn-cross:, and :btn-square: but no :btn-a: or
> :btn-b:.)
```

...then to these two lines:

```markdown
> Rather than being part of the code, the addition of :btn-start-nes: and/or
> :btn-select-nes: is actually part of selecting the number of players and
```

...and finally to to this test case:

```rust
#[test]
fn pulldown_cmark_0_12_1_deflist_panic() {
    let markdown = "> foo\n> :word-wrapped-shortcode:";
    let mut options = pulldown_cmark::Options::empty();
    options.insert(pulldown_cmark::Options::ENABLE_DEFINITION_LIST);

    Parser::new_ext(markdown, options);
}
```

Here's the traceback:

```
---- markdown::test::pulldown_cmark_0_12_1_deflist_panic stdout ----
thread 'markdown::test::pulldown_cmark_0_12_1_deflist_panic' panicked at /home/ssokolow/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.1/src/firstpass.rs:1507:38:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panicking.rs:146:5
   3: core::option::unwrap_failed
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/option.rs:1985:5
   4: core::option::Option<T>::unwrap
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/option.rs:935:21
   5: pulldown_cmark::firstpass::FirstPass::pop
             at /home/ssokolow/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.1/src/firstpass.rs:1507:38
   6: pulldown_cmark::firstpass::FirstPass::run
             at /home/ssokolow/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.1/src/firstpass.rs:79:13
   7: pulldown_cmark::firstpass::run_first_pass
             at /home/ssokolow/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.1/src/firstpass.rs:41:5
   8: pulldown_cmark::parse::Parser<F>::new_with_broken_link_callback
             at /home/ssokolow/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.1/src/parse.rs:247:34
   9: pulldown_cmark::parse::Parser::new_ext
             at /home/ssokolow/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.1/src/parse.rs:232:9
  10: render_wishlist::markdown::test::pulldown_cmark_0_12_1_deflist_panic
             at ./src/markdown.rs:563:9
  11: render_wishlist::markdown::test::pulldown_cmark_0_12_1_deflist_panic::{{closure}}
             at ./src/markdown.rs:558:45
  12: core::ops::function::FnOnce::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:250:5
  13: core::ops::function::FnOnce::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

(Yeah. `render_wishlist` is the original purpose of the project and I haven't renamed it for its broader use yet.)

Basically, it looks like there's some kind of "Once you start something that might be definition list markup, you'd better jolly well finish it as valid definition list markup or we're gonna crash" bug... but, funny enough, so far, I've only managed to get it to trigger when the potential DL markup is inside a blockquote.

**Comment by thscharler** [comment]
I might have encountered the same problem. 
My version is pulldown-cmark v0.12.0

```
* def this
  : def text def text
```

gives a panic too. 


```
thread 'main' panicked at C:\Users\stommy\.cargo\registry\src\index.crates.io-6f17d22bba15001f\pulldown-cmark-0.12.0\src\firstpass.rs:1507:38:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\std\src\panicking.rs:652
   1: core::panicking::panic_fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\core\src\panicking.rs:72
   2: core::panicking::panic
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\core\src\panicking.rs:146
   3: core::option::unwrap_failed
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\core\src\option.rs:1985
   4: enum2$<core::option::Option<pulldown_cmark::tree::TreeIndex> >::unwrap
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\core\src\option.rs:935
   5: pulldown_cmark::firstpass::FirstPass::pop
             at C:\Users\stommy\.cargo\registry\src\index.crates.io-6f17d22bba15001f\pulldown-cmark-0.12.0\src\firstpass.rs:1507
   6: pulldown_cmark::firstpass::FirstPass::run
             at C:\Users\stommy\.cargo\registry\src\index.crates.io-6f17d22bba15001f\pulldown-cmark-0.12.0\src\firstpass.rs:79
   7: pulldown_cmark::firstpass::run_first_pass
             at C:\Users\stommy\.cargo\registry\src\index.crates.io-6f17d22bba15001f\pulldown-cmark-0.12.0\src\firstpass.rs:41
   8: pulldown_cmark::parse::Parser<pulldown_cmark::parse::DefaultBrokenLinkCallback>::new_with_broken_link_callback<pulldown_cmark::parse::DefaultBrokenLinkCallback>
             at C:\Users\stommy\.cargo\registry\src\index.crates.io-6f17d22bba15001f\pulldown-cmark-0.12.0\src\parse.rs:247
   9: pulldown_cmark::parse::Parser<pulldown_cmark::parse::DefaultBrokenLinkCallback>::new_ext
             at C:\Users\stommy\.cargo\registry\src\index.crates.io-6f17d22bba15001f\pulldown-cmark-0.12.0\src\parse.rs:232
  10: mdedit::mdedit_parts::markdown::dump_md
             at .\rat-salsa\examples\mdedit_parts\markdown.rs:1899
```

**Comment by ssokolow** [comment]
I'm still getting this panic when I try enabling definition list support after updating to v0.12.2, despite it listing #954 in its release notes.

```
thread 'main' panicked at /home/ssokolow/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pulldown-cmark-0.12.2/src/firstpass.rs:1540:38:
called `Option::unwrap()` on a `None` value
```

Should I file a new issue?

**Comment by ollpu** [comment]
The release doesn't include #974 which might fix further issues.

@ssokolow can you provide an input that still crashes so we can test it?

**Comment by ssokolow** [comment]
> @ssokolow can you provide an input that still crashes so we can test it?

```rust
#[test]
fn pulldown_cmark_0_12_2_deflist_panic() {
    let mut options = pulldown_cmark::Options::empty();
    options.insert(pulldown_cmark::Options::ENABLE_DEFINITION_LIST);
    Parser::new_ext("> x\n> :xx:\n\nx", options);
}
```

**Comment by ollpu** [comment]
@ssokolow Thanks! That seems to work on master, and I'm not finding any more crashes with fuzzing.

I suppose we can do another minor release at some point?

**Comment by Martin1887** [comment]
Uhm... I think #978 is more a fix than a breaking change, but it is in the middle.

**Comment by Martin1887** [comment]
I guess this is fixed in the latest release 🙂.
